File: Factory method/python/FactoryMethod.py
class IProduct:

    def release(self):
        pass


class Car(IProduct):

    def release(self):
        print('car')


class Truck(IProduct):

    def release(self):
        print('truck')


class IWorkShop:

    def create(self) -> IProduct:
        pass


class CarWorkShop(IWorkShop):

    def create(self) -> IProduct:
        return Car()


class TruckWorkShop(IWorkShop):

    def create(self) -> IProduct:
        return Truck()


if __name__ == '__main__':

    creator = CarWorkShop()
    car = creator.create()

    creator = TruckWorkShop()
    truck = creator.create()

    car.release()
    truck.release()

File: Factory method/c++/FactoryMethod.cpp
#include <iostream>

class IProduction
{
    public:
        virtual void release() = 0;
};

class Car : public IProduction
{
    public:
        void release() override;
};

class Truck : public IProduction
{
    public:
        void release() override;
};

class IWorkShop
{
    public:
        virtual IProduction* create() = 0;
};

class CarWorkShop : public IWorkShop
{
    public:
        IProduction* create() override;
};

class TruckWorkShop : public IWorkShop
{
    public:
        IProduction* create() override;
};

int main()
{
    IWorkShop* creator = new CarWorkShop();

    IProduction* car = creator->create();

    creator = new TruckWorkShop();

    IProduction* truck = creator->create();

    car->release();
    truck->release();

    return 0;
}

void Car::release()
{
    std::cout << "car\n";
}

void Truck::release()
{
    std::cout << "truck\n";
}

IProduction *CarWorkShop::create()
{
    return new Car();
}

IProduction *TruckWorkShop::create()
{
    return new Truck();
}